ENH Allows plotting max class for multiclass in `DecisionBoundaryDisplay`

lucyleeow opened the pull request:
#### Reference Issues/PRs
Towards #27462

#### What does this implement/fix? Explain your changes.
Allows `DecisionBoundaryDisplay` to represent all classes for multiclass `decision_function` and `predict_probas` by plotting the class with the max response at each point.

Have closely followed the code here: https://github.com/scikit-learn/scikit-learn/pull/27291#issuecomment-1708716064

Not 100% on what the colour API should be, open to change.

`contour` and `contourf` have a `colors` parameter, which can be used instead of `cmap` BUT `pcolormesh` only has `cmap`. For simplicity, I've decided to only allow users to pass `cmap`. This probably makes more sense in this context vs a list of colors to cycle through.

#### Any other comments?
WIP - need to add tests once we're happy with plot appearance and API

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lucyleeow committed:
wip, tests pass, no new tests yet

lucyleeow committed:
lint

lucyleeow committed:
add whats new

lucyleeow committed:
format

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -254,18 +292,32 @@ def from_estimator(
 
             .. versionadded:: 1.4
 
+        multiclass_cmap : list[str, Colormap] default=None
```
Now that I'm looking at this, I'm thinking that the best API would be a list of strings containing primary colors:

```python
DecisionBoundaryDisplay.from_estimator(model, X, colors=["tab:blue", "tab:orange", "tab:green"])
```

I would probably rename the parameter to `colors` or `class_colors` if there is an ambiguity.
We can get the RGB representation of those standard Matplotlib color with

```python
import matplotlib as mpl
r, g, b = mpl.colors.to_rgb("tab:blue")
```

Then, the idea will be to always a `LinearSegmentedColormap` from each color.

glemaitre replied:
Basically, I think that we cover the major use cases and we can point to this tutorial: https://matplotlib.org/stable/gallery/color/named_colors.html to get the name of the colors.

glemaitre replied:
So as a short example, it would like:

```python
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression
from sklearn.inspection import DecisionBoundaryDisplay

X, y = load_iris(return_X_y=True)
X = X[:, :2]

model = LogisticRegression().fit(X, y)
disp = DecisionBoundaryDisplay.from_estimator(
    model, X, multiclass_cmap=["r", "g", "b"]
)

color_map = {0: 'r', 1: 'g', 2: 'b'}
colors = [color_map[i] for i in y]
disp.ax_.scatter(X[:, 0], X[:, 1], c=colors)
```

![image](https://github.com/user-attachments/assets/f8fd2c44-5707-4edd-befb-e39bf136e7fc)

lucyleeow replied:
I think this idea would be better but let me clarify what you mean.

> I would probably rename the parameter to colors or class_colors if there is an ambiguity.

Would this parameter only be used when plotting all classes in multi-class case? If that is the case, why not use the term 'multiclass' in the parameter?

If that is not the case, I think this could be confusing. Both `contour` and `contourf` have `colors` AND `cmap` parameters (only one should be set). If we want to use this term outside of the multiclass case, it would be difficult to determine whether to pass the arg to `colors` or `cmap` ?

---

I do think this idea is better because when plotting several classes, you would generally want to use a single color colormap per class, otherwise it would be visually confusing. It also may be easier as color names may be more well known than cmap names.

The cons:

* May be a little confusing as we are asking for 'colors' but we will actually be passing these on to `cmap`
* User will not be able to use their own colormap or a registered mpl cmap.

glemaitre replied:
> If that is the case, why not use the term 'multiclass' in the parameter?

I was more against using the term `cmap` indeed. I would be fine with `multiclass` appearing.

glemaitre replied:
> User will not be able to use their own colormap or a registered mpl cmap.

Thinking a bit more, we could accept registered colormap from matplotlib and do the same process that we do for `viridis`.

So the option would be a list of colors that should match `n_classes` or a colormap with a single string that we will resample to get `n_classes` colors.

glemaitre replied:
So we will need a name to reflect both. I assume that we might want to get away from `color` or `cmap` and a name that reflect better something about our display.

lucyleeow replied:
Okay I've amended to list of colors or cmap str. Let me know what you think.

I named the parameter 'multiclass_colors', couldn't think of a better name and I feel that we sort of need the word 'color' because that's what we are doing... but I am open to change though!

(I will add tests etc once we are happy with API)

glemaitre replied:
Let's start with this API. We will see how it goes with the second reviewer. In the worst case scenario, it will come back to a search and replace.

glemaitre replied:
I'm resolving this discussion then.

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -343,6 +395,18 @@ def from_estimator(
+            response_method in ("predict_proba", "decision_function", "auto")
+            and multiclass_cmap is not None
+            and hasattr(estimator, "classes_")
+            and (n_classes := len(estimator.classes_) > 2)
```
I think there is a bug here:

```suggestion
            and (n_classes := len(estimator.classes_)) > 2
```

lucyleeow committed:
review

lucyleeow committed:
use color list or cmap str

lucyleeow committed:
wording

lucyleeow committed:
cruft

glemaitre commented on doc/whats_new/v1.6.rst:
```diff
@@ -236,6 +236,14 @@ Changelog
+
+- |Enhancement| :class:`inspection.DecisionBoundaryDisplay` now supports
+  plotting all classes for multi-class problems when `response_method` is
+  'decision_function' or 'predict_proba'.
```
as well `"auto"`

lucyleeow committed:
review

lucyleeow committed:
Merge branch 'main' into dbplot_multi

lucyleeow committed:
add tests

lucyleeow committed:
Merge branch 'main' into dbplot_multi

lucyleeow commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -191,7 +209,45 @@ def plot(self, plot_method="contourf", ax=None, xlabel=None, ylabel=None, **kwar
+                )
+                # `cmap` should not be in kwargs
+                if "cmap" in kwargs:
+                    del kwargs["cmap"]
```
Not 100% I should be deleting a `kwarg` here...

glemaitre replied:
Yep this might be safer to raise an error instead mentioning that you cannot pass a `cmap` parameter and instead pass `multiclass_colors`.

glemaitre replied:
Otherwise we could create a new dictionary that just remove cmap and raise the warning but you are right that we should not modify the original dictionary.

lucyleeow replied:
I make a copy of the dict, `safe_kwargs`. Happy to change though!

lucyleeow commented:
I'm also going to try amend example: [Logistic Regression 3-class Classifier](https://scikit-learn.org/dev/auto_examples/linear_model/plot_iris_logistic.html#sphx-glr-auto-examples-linear-model-plot-iris-logistic-py) to show a plot using this enhancement. Though as we use the same `X` we used to fit the estimator, the plot probably won't look great.
I could instead use `Nystroem`, though this would only for the purpose of showing a prettier plot...WDYT @glemaitre ?

lucyleeow committed:
lint

glemaitre commented:
I would almost argue that we be removing the `LogisticRegression` plot because it does not have that much values :).

I'm thinking that we could instead modify this one: https://scikit-learn.org/stable/auto_examples/classification/plot_classification_probability.html#sphx-glr-auto-examples-classification-plot-classification-probability-py and add an additional column to the plot with the max. Like this we could remove the OvR LR (it is deprecated), and instead use a Nystroem with LR.

glemaitre reviewed (commented):
Ups I see I didn't post some review that I had. This should be partial comments.

lucyleeow commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -88,9 +79,23 @@ class DecisionBoundaryDisplay:
 
+    multiclass_colors : list[str], str, default=None
+        Specifies how to color each class when plotting all classes of multiclass
+        problem and `response.shape.ndim==3`. Ignored in all other cases.
```
Yeah I can make it more user friendly.

The other case is binary or multi-class but `class_of_interest` is not None. So its a bit wordy but I should make it more user friendly.

glemaitre replied:
Indeed, I think this is worth writing the full statement.

lucyleeow commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -191,7 +207,37 @@ def plot(self, plot_method="contourf", ax=None, xlabel=None, ylabel=None, **kwar
+        if self.response.ndim == 2:
+            self.surface_ = plot_func(self.xx0, self.xx1, self.response, **kwargs)
+        else:  # self.response.ndim == 3
+            colors = self.multiclass_colors
```
Done (it was just long)

lucyleeow committed:
review

lucyleeow commented:
> I'm thinking that we could instead modify this one: https://scikit-learn.org/stable/auto_examples/classification/plot_classification_probability.html#sphx-glr-auto-examples-classification-plot-classification-probability-py and add an additional column to the plot with the max. Like this we could remove the OvR LR (it is deprecated), and instead use a Nystroem with LR.

Do you want me to do this in this PR or a separate PR (since removing OvR LR may *technically* not be considered part of this work)?

lucyleeow committed:
docstring format

lucyleeow committed:
fix kwargs

glemaitre commented:
> Do you want me to do this in this PR or a separate PR (since removing OvR LR may technically not be considered part of this work)?

We can do it in this PR and limit the change to this single example.

lucyleeow committed:
Merge branch 'main' into dbplot_multi

lucyleeow committed:
fix for old mpl

lucyleeow committed:
fix for old mpl

lucyleeow committed:
fix for old mpl

lucyleeow committed:
check

lucyleeow committed:
lint

lucyleeow committed:
try

glemaitre commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,16 +575,85 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+        qq = quad.cmap._segmentdata
+        testcc = cmaps[idx]._segmentdata
+        assert (
+            quad.cmap == cmaps[idx]
```
If this is failing, we can probably check some attributes such as the name, and the limit (`.get_under()` and `_get_over()`). I'll check if there is any other meaningful attribute.

glemaitre replied:
The attribute `N` might also be useful. Basically, with the lower bound higher bound and the level of quantization, we are almost certain that we will get the same colormap.

lucyleeow replied:
So I think this is failing because in mpl min version (3.3.4), `__eq__` is not implemented for `LinearSegmentedColormap`. Trying to find the PR or doc for this, so I can make a good comment in the code but not successful.

I think I will just skip for older mpl version, WDYT?

lucyleeow replied:
Ah okay this was added in 3.5 in `colormaps` class, so it will check for equality even when its not the same object, see: https://github.com/matplotlib/matplotlib/pull/20227

glemaitre replied:
Yep that's why I thought that we could check the couple of public of attributes allowing to check that everything is as expected: name and bounds

lucyleeow replied:
Ah yup, I can just manually test what `__eq__` checks. Will amend!

lucyleeow replied:
I've just checked that `_lut` is equal, which seems to be the main think that is checked in the new `__eq__` method of `Colormaps` in mpl. Happy to change/add.

lucyleeow replied:
Not 100% that using the private `_lut` is the best option (note need to init first). I do think this is the best item to compare (as it is what is done in `__eq__` method, and e.g., `get_under`() etc do use `._lut`, e.g.:

```python
    def get_under(self):
        """Get the color for low out-of-range values."""
        if not self._isinit:
            self._init()
        return np.array(self._lut[self._i_under])
```

Happy to change.

glemaitre commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,16 +575,85 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+)
+@pytest.mark.parametrize("plot_method", ["contourf", "contour", "pcolormesh"])
+def test_multiclass_colors_cmap(pyplot, plot_method, multiclass_colors):
+    """Check correct cmap used  for all `multiclass_colors` inputs."""
```
```suggestion
    """Check correct cmap used for all `multiclass_colors` inputs."""
```

glemaitre commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,16 +575,85 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+        multiclass_colors=multiclass_colors,
+    )
+
+    cmaps = []
+    if multiclass_colors == "plasma":
+        cmap = mpl.pyplot.get_cmap(multiclass_colors, len(clf.classes_))
+        for class_idx, primary_color in enumerate(cmap.colors):
+            r, g, b, _ = primary_color
+            class_cmap = mpl.colors.LinearSegmentedColormap.from_list(
+                f"colormap_{class_idx}", [(1.0, 1.0, 1.0, 1.0), (r, g, b, 1.0)]
+            )
+            cmaps.append(class_cmap)
+    else:
+        for class_idx, class_color in enumerate(multiclass_colors):
+            r, g, b, _ = mpl.colors.to_rgba(class_color)
+            class_cmap = mpl.colors.LinearSegmentedColormap.from_list(
+                f"colormap_{class_idx}", [(1.0, 1.0, 1.0, 1.0), (r, g, b, 1.0)]
+            )
+            cmaps.append(class_cmap)
```
I think that we can refactor with the following:

```suggestion
    cmaps = []
    if multiclass_colors == "plasma":
        colors = mpl.pyplot.get_cmap(multiclass_colors, len(clf.classes_)).colors
    else:
        colors = [mpl.colors.to_rgba(color) for color in multiclass_colors]

    for class_idx, (r, g, b, _) in enumerate(colors):
        class_cmap = mpl.colors.LinearSegmentedColormap.from_list(
            f"colormap_{class_idx}", [(1.0, 1.0, 1.0, 1.0), (r, g, b, 1.0)]
        )
        cmaps.append(class_cmap)
```

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -191,7 +207,51 @@ def plot(self, plot_method="contourf", ax=None, xlabel=None, ylabel=None, **kwar
 
         plot_func = getattr(ax, plot_method)
-        self.surface_ = plot_func(self.xx0, self.xx1, self.response, **kwargs)
+        if self.response.ndim == 2:
+            self.surface_ = plot_func(self.xx0, self.xx1, self.response, **kwargs)
+        else:  # self.response.ndim == 3
+            multiclass_cmaps = []
+            if (
+                isinstance(self.multiclass_colors, str)
+                or self.multiclass_colors is None
+            ):
+                cmap = (
+                    "viridis"
+                    if self.multiclass_colors is None
+                    else self.multiclass_colors
+                )
+                cmap = plt.get_cmap(cmap, self.response.shape[-1])
+                for class_idx, (r, g, b, _) in enumerate(cmap.colors):
+                    class_cmap = mpl.colors.LinearSegmentedColormap.from_list(
+                        f"colormap_{class_idx}", [(1.0, 1.0, 1.0, 1.0), (r, g, b, 1.0)]
+                    )
+                    multiclass_cmaps.append(class_cmap)
+            else:
+                for class_idx, class_color in enumerate(self.multiclass_colors):
+                    r, g, b, _ = mpl.colors.to_rgba(class_color)
+                    class_cmap = mpl.colors.LinearSegmentedColormap.from_list(
+                        f"colormap_{class_idx}", [(1.0, 1.0, 1.0, 1.0), (r, g, b, 1.0)]
+                    )
+                    multiclass_cmaps.append(class_cmap)
```
I think that we can simplify also in this part of the code

```suggestion
        if self.response.ndim == 2:
            self.surface_ = plot_func(self.xx0, self.xx1, self.response, **kwargs)
        else:  # self.response.ndim == 3
            multiclass_cmaps = []
            if isinstance(self.multiclass_colors, str) or self.multiclass_colors is None:
                cmap = "viridis" if self.multiclass_colors is None else self.multiclass_colors
                cmap = plt.get_cmap(cmap, self.response.shape[-1])
                colors = cmap.colors
            else:
                colors = [mpl.colors.to_rgba(color) for color in self.multiclass_colors]

            for class_idx, (r, g, b, _) in enumerate(colors):
                class_cmap = mpl.colors.LinearSegmentedColormap.from_list(
                    f"colormap_{class_idx}", [(1.0, 1.0, 1.0, 1.0), (r, g, b, 1.0)]
                )
                multiclass_cmaps.append(class_cmap)
```

lucyleeow committed:
skip old mpl

lucyleeow committed:
review

lucyleeow committed:
check array instead

lucyleeow committed:
lint

lucyleeow committed:
Merge branch 'main' into dbplot_multi

lucyleeow committed:
init first

lucyleeow committed:
add name check

lucyleeow committed:
lint

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -254,18 +312,39 @@ def from_estimator(
 
             .. versionadded:: 1.4
 
+        multiclass_colors : list of str, or str, default=None
```
I would change as well this part then.

glemaitre reviewed (commented):
A partial review on the code. I'll check the tests soonish.

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -191,7 +207,48 @@ def plot(self, plot_method="contourf", ax=None, xlabel=None, ylabel=None, **kwar
+                    if self.multiclass_colors is None
+                    else self.multiclass_colors
+                )
+                cmap = plt.get_cmap(cmap, self.response.shape[-1])
```
For instance, the error will be already be good:

```pytb
In [6]: plt.get_cmap("xxx", 10)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[6], line 1
----> 1 plt.get_cmap("xxx", 10)

File ~/mambaforge/envs/sklearn_dev/lib/python3.10/site-packages/matplotlib/pyplot.py:2563, in get_cmap(name, lut)
   2561 if isinstance(name, Colormap):
   2562     return name
-> 2563 _api.check_in_list(sorted(_colormaps), name=name)
   2564 if lut is None:
   2565     return _colormaps[name]

File ~/mambaforge/envs/sklearn_dev/lib/python3.10/site-packages/matplotlib/_api/__init__.py:129, in check_in_list(values, _print_supported_values, **kwargs)
    127 if _print_supported_values:
    128     msg += f"; supported values are {', '.join(map(repr, values))}"
--> 129 raise ValueError(msg)

ValueError: 'xxx' is not a valid value for name; supported values are 'Accent', 'Accent_r', 'Blues', 'Blues_r', 'BrBG', 'BrBG_r', 'BuGn', 'BuGn_r', 'BuPu', 'BuPu_r', 'CMRmap', 'CMRmap_r', 'Dark2', 'Dark2_r', 'GnBu', 'GnBu_r', 'Grays', 'Greens', 'Greens_r', 'Greys', 'Greys_r', 'OrRd', 'OrRd_r', 'Oranges', 'Oranges_r', 'PRGn', 'PRGn_r', 'Paired', 'Paired_r', 'Pastel1', 'Pastel1_r', 'Pastel2', 'Pastel2_r', 'PiYG', 'PiYG_r', 'PuBu', 'PuBuGn', 'PuBuGn_r', 'PuBu_r', 'PuOr', 'PuOr_r', 'PuRd', 'PuRd_r', 'Purples', 'Purples_r', 'RdBu', 'RdBu_r', 'RdGy', 'RdGy_r', 'RdPu', 'RdPu_r', 'RdYlBu', 'RdYlBu_r', 'RdYlGn', 'RdYlGn_r', 'Reds', 'Reds_r', 'Set1', 'Set1_r', 'Set2', 'Set2_r', 'Set3', 'Set3_r', 'Spectral', 'Spectral_r', 'Wistia', 'Wistia_r', 'YlGn', 'YlGnBu', 'YlGnBu_r', 'YlGn_r', 'YlOrBr', 'YlOrBr_r', 'YlOrRd', 'YlOrRd_r', 'afmhot', 'afmhot_r', 'autumn', 'autumn_r', 'binary', 'binary_r', 'bone', 'bone_r', 'brg', 'brg_r', 'bwr', 'bwr_r', 'cividis', 'cividis_r', 'cool', 'cool_r', 'coolwarm', 'coolwarm_r', 'copper', 'copper_r', 'cubehelix', 'cubehelix_r', 'flag', 'flag_r', 'gist_earth', 'gist_earth_r', 'gist_gray', 'gist_gray_r', 'gist_grey', 'gist_heat', 'gist_heat_r', 'gist_ncar', 'gist_ncar_r', 'gist_rainbow', 'gist_rainbow_r', 'gist_stern', 'gist_stern_r', 'gist_yarg', 'gist_yarg_r', 'gist_yerg', 'gnuplot', 'gnuplot2', 'gnuplot2_r', 'gnuplot_r', 'gray', 'gray_r', 'grey', 'hot', 'hot_r', 'hsv', 'hsv_r', 'inferno', 'inferno_r', 'jet', 'jet_r', 'magma', 'magma_r', 'nipy_spectral', 'nipy_spectral_r', 'ocean', 'ocean_r', 'pink', 'pink_r', 'plasma', 'plasma_r', 'prism', 'prism_r', 'rainbow', 'rainbow_r', 'seismic', 'seismic_r', 'spring', 'spring_r', 'summer', 'summer_r', 'tab10', 'tab10_r', 'tab20', 'tab20_r', 'tab20b', 'tab20b_r', 'tab20c', 'tab20c_r', 'terrain', 'terrain_r', 'turbo', 'turbo_r', 'twilight', 'twilight_r', 'twilight_shifted', 'twilight_shifted_r', 'viridis', 'viridis_r', 'winter', 'winter_r'
```

glemaitre replied:
And also add a test for this case.

glemaitre commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -191,7 +207,48 @@ def plot(self, plot_method="contourf", ax=None, xlabel=None, ylabel=None, **kwar
+                cmap = plt.get_cmap(cmap, self.response.shape[-1])
+                colors = cmap.colors
+            else:
+                colors = [mpl.colors.to_rgba(color) for color in self.multiclass_colors]
```
So we should also add a test for this case

lucyleeow replied:
Done in a slightly different way in: https://github.com/lucyleeow/scikit-learn/blob/c58d656e61ec70ef7b34a726e5b6d21d4e7266ed/sklearn/inspection/_plot/decision_boundary.py#L441

Also check cmap is correct too. Also have test for both. Happy to change though!

glemaitre reviewed (commented):
It looks good (you can discard some of my previous comments; apparently I did not pull before commenting).

I think that we can improve a bit the test.

lucyleeow commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,16 +575,81 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+    clf = LogisticRegression().fit(X, y)
+
+    disp = DecisionBoundaryDisplay.from_estimator(
+        clf,
+        X,
+        response_method=response_method,
```
AH yes, I like to add a trailing comma but when black sees this it puts everything on separate lines :grimacing: will change!

lucyleeow committed:
merge main

lucyleeow commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -88,9 +81,28 @@ class DecisionBoundaryDisplay:
+        single prediction value per point.
+        Possible inputs:
+
+        * list, of length `response.shape[-1]`, of Matplotlib `color
```
> Instead of response.shape[-1] can we just say n_classes or there is a corner case that I'm not thinking of?

So I've been weirdly avoiding the use of e.g., `n_classes` when the estimator does not have that attribute/parameter (because it's in backticks and 'looks' like a parameter? but honestly just neurotic-ness). But I will just use it, because I think users know what this means.

lucyleeow commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,16 +575,81 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+
+    for idx, quad in enumerate(disp.surface_):
+        # Ensure `_lut` (look-up table) is equal
+        quad.cmap._init()
+        cmaps[idx]._init()
+        assert_array_equal(quad.cmap._lut, cmaps[idx]._lut)
```
Okay I think I will use the `==` for mpl versions >=3.5 and check for bounds for lower versions. I will add a TODO to remove the latter check once our min mpl version is 3.5. I think this would be ideal, as `==` is what we ideally want to use.

lucyleeow committed:
review

glemaitre commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,17 +576,83 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+        for class_idx, (r, g, b, _) in enumerate(colors)
+    ]
+
+    for idx, quad in enumerate(disp.surface_):
+        if parse_version(mpl.__version__) >= parse_version("3.5"):
+            assert quad.cmap == cmaps[idx]
+        # TODO: remove once min mpl version >=3.5
+        else:
+            # Ensure lut (look-up table) bounds are equal
+            assert_array_equal(quad.cmap.get_over(), cmaps[idx].get_over())
+            assert_array_equal(quad.cmap.get_under(), cmaps[idx].get_under())
+            # Ensure name is the same
+            assert quad.cmap.name == cmaps[idx].name
```
actually, we could execute this check only for when matplotlib implemented the `__eq__`. I think that this is enough.

lucyleeow committed:
fix doc?

lucyleeow committed:
merge main

lucyleeow committed:
add whats new

lucyleeow committed:
try fix doc

lucyleeow committed:
skip for mpl < 3.5

lucyleeow committed:
fix mpl version skip

lucyleeow committed:
improve test

lucyleeow committed:
lint

lucyleeow commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,16 +575,81 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+
+    grid = np.concatenate([disp.xx0.reshape(-1, 1), disp.xx1.reshape(-1, 1)], axis=1)
+    response = getattr(clf, response_method)(grid)
+    assert_allclose(response.reshape(*disp.response.shape), disp.response)
```
You're completely right, I don't really test anything atm. Done, and I've used `pcolormesh` because it was much easier to test as the other two methods only gave you the contour lines.

lucyleeow committed:
try fix for old mpl

lucyleeow committed:
revert color error catch

glemaitre commented:
Uhm it looks like the CI fails.

lucyleeow committed:
fix test

lucyleeow commented on sklearn/inspection/_plot/tests/test_boundary_decision_display.py:
```diff
@@ -577,18 +576,97 @@ def test_class_of_interest_multiclass(pyplot, response_method):
+    """Check plot correct when plotting max multiclass class."""
+    import matplotlib as mpl
+
+    # In matplotlib < v3.5, default value of `pcolormesh(shading)` is 'flat',
+    # which results in the last row and column of the color mapped values being
+    # dropped. This results in a 99x99 grid.
+    if parse_version(mpl.__version__) < parse_version("3.5"):
+        pytest.skip(
+            "`pcolormesh` in Matplotlib >= 3.5 gives smaller grid size."
```
So min mpl version problem again.

Alternatively, I can add an if/else and for mpl < 3.5 and amend the test to something like:

```python
highest_class_mask = highest_class != idx
assert_array_equal(highest_class_mask[:-1,:-1].flatten(), quadmesh.get_array().mask)
```

and I can have a comment to delete this once we bump the min version of mpl (currently 3.3.4)

lucyleeow committed:
lint

lucyleeow committed:
fix test

glemaitre commented on doc/whats_new/upcoming_changes/sklearn.inspection/29797.enhancement.rst:
```diff
@@ -0,0 +1,4 @@
+- |Enhancement| :class:`inspection.DecisionBoundaryDisplay` now supports
```
```suggestion
- :class:`inspection.DecisionBoundaryDisplay` now supports
```

glemaitre commented on doc/whats_new/upcoming_changes/sklearn.inspection/29797.enhancement.rst:
```diff
@@ -0,0 +1,4 @@
+- |Enhancement| :class:`inspection.DecisionBoundaryDisplay` now supports
+  plotting all classes for multi-class problems when `response_method` is
+  'decision_function', 'predict_proba' or 'auto'.
+  By :user:`Lucy Liu <lucyleeow>`.
```
```suggestion
  By :user:`Lucy Liu <lucyleeow>`
```

lucyleeow committed:
review

lucyleeow commented:
Here is an attempt at adding a column to the plot in the [Plot classification probability](https://scikit-learn.org/dev/auto_examples/classification/plot_classification_probability.html#sphx-glr-auto-examples-classification-plot-classification-probability-py):

![image](https://github.com/user-attachments/assets/d9ee049a-0ce2-49bf-ba0d-ce5e3495270a)

I played around with different colormaps and thought that a single color colormap for the single class plots was better than having two different diverging colormaps.

I can also move the colorbars so that they are to the left and right of the plots.

WDYT @glemaitre ?

glemaitre commented:
It looks good. I would be OK keeping the colormap at the location you have them now. I would only increase the width of the figure.

Would it be to much to have the scatter plot as well on the last plot (last column)?

lucyleeow committed:
wip example update

lucyleeow committed:
amend example

lucyleeow committed:
lint

lucyleeow committed:
try inc figsize

lucyleeow committed:
fix figure

lucyleeow committed:
lint

lucyleeow commented:
Okay I think this is ready to review again @glemaitre. I have updated the example `plot_classification_probability.py`, some notes:

* I removed the name 'multinomial' on because both the first 2 classifiers are multinomial (and now also the 3rd)
* I think it is obvious in the last column plots, which class the scatter plot points belong to (i.e. purple points = class 0). I thought about adding a legend, but I thought it would add unnecessary complexity to the plot.
* Should we add any comment on RBF kernel + logistic plot, re: class 2 prediction at the corners of the plot (though I don't know exactly why this would be, I would guess it's because the higher dim transformation of features..?)

glemaitre commented:
Cool. I will have a look now.

glemaitre commented on examples/classification/plot_classification_probability.py:
```diff
@@ -4,22 +4,27 @@
-The logistic regression with One-Vs-Rest is not a multiclass classifier out of
-the box. As a result it has more trouble in separating class 2 and 3 than the
-other estimators.
+The first three columns shows the predicted probability for varing values of the two
```
```suggestion
The first three columns shows the predicted probability for varying values of the two
```

glemaitre reviewed (approved):
LGTM

lucyleeow committed:
review

lucyleeow committed:
Merge branch 'main' into dbplot_multi

lucyleeow committed:
Merge branch 'main' into dbplot_multi

lucyleeow commented:
@ogrisel gentle ping about this :grimacing: Thank you!

ogrisel committed:
Use tab10 cmap by default and adjust to the number of classes

ogrisel committed:
Refactor / enrich plot_classification_probability.py

ogrisel commented:
@lucyleeow I pushed two commits of changes I wanted to do while reviewing the PR:

- the first selects a qualitative colormap automatically and adjusts it to the observed number of classes;
- the second is a refactoring of the example to give an example on how to read those plots to draw conclusions on the impact of models parameters on the shape of the decision boundary and the magnitude of the predicted probabilities.

I will do a full review tomorrow but this LGTM.

ogrisel commented on examples/classification/plot_classification_probability.py:
```diff
@@ -3,93 +3,240 @@
             classifier,
-            X,
+            X_train,
             response_method="predict_proba",
```
This is a pitfall of this method. By default, it would use `decision_function` which is much harder to interpret in my opinion (especially when comparing different model classes).

I think we should change the "auto" policy to use favor `predict_proba` when available but this should rather be done in a dedicated follow-up PR.

ogrisel replied:
BTW, if we use `response_method="predict_proba"`, maybe the `vmin=0` and `vmax=1` parameters could be set automatically to make the use `DecisionBoundaryDisplay` terser.

glemaitre replied:
It looks indeed like a good suggestion.

ogrisel commented on sklearn/inspection/_plot/decision_boundary.py:
```diff
@@ -88,9 +81,28 @@ class DecisionBoundaryDisplay:
+        colors taken from the colormap and passed to the `cmap` parameter of
+        the `plot_method`.
+
+        .. versionadded:: 1.6
```
```suggestion
        .. versionadded:: 1.7
```

ogrisel committed:
Update versionadded

lucyleeow commented:
FYI @ogrisel I'm just going to fix the merge conflicts

lucyleeow committed:
merge main

ogrisel committed:
Grammar and split long sentences

ogrisel committed:
Pass quantile_method='averaged_inverted_cdf' to KBinsDiscretize to avoid raising warning

ogrisel commented:
Thanks @lucyleeow!

lucyleeow commented:
Thanks for finishing this one off @ogrisel !

glemaitre commented:
Thanks @ogrisel

ogrisel commented:
@lucyleeow If you are looking for follow-up improvements, please consider those two comments:

https://github.com/scikit-learn/scikit-learn/pull/29797#discussion_r1983772956

lesteve commented:
FYI this broke the CI when matplotlib is not installed, fix in https://github.com/scikit-learn/scikit-learn/pull/30971
